### ISSUE TITLE
How can i remove the YOLOv5 Detect() when convert ckpt to onnx?

### ISSUE THREAD

**Issue by rikichou** [ultralytics/yolov5#3077]
## ❔Question
How can i remove the YOLOv5 Detect() when convert ckpt to onnx?

## Additional context
I fllow the readme to export onnx:
python models/export.py --weights runs\train\exp3\weights\best.pt --img 640 --batch 1

and the output best.onnx also include Detection head.

and I add model.model[-1].export = True to the export.py, but it is no use

How can i remove the YOLOv5 Detect() when convert ckpt to onnx?

**Comment by glenn-jocher** [comment]
@rikichou if you would like to remove the detection head from any YOLOv5 model you can simply remove the last layer:
```
model.model = model.model[:-1]
```

**Comment by rikichou** [comment]
> @rikichou if you would like to remove the detection head from any YOLOv5 model you can simply remove the last layer:
> 
> ```
> model.model = model.model[:-1]
> ```

thanks, That's work!!!!

**Comment by rikichou** [comment]
if you wanna use ncnn's yolov5 demo cpp, do not add model.model = model.model[:-1].

just fix export.py   (add training=TrainingMode.TRAINING for torch.onnx.export)

         print(f'{prefix} starting export with onnx {onnx.__version__}...')
         f = opt.weights.replace('.pt', '.onnx')  # filename
+        torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
                           dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # size(1,3,640,640)
                                         'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)

**Comment by YuhangLiu-602** [comment]
![截屏2021-05-08 下午9 06 19](https://user-images.githubusercontent.com/65380823/117540361-97bc6080-b041-11eb-85a3-296c257d9f46.png)
![截屏2021-05-08 下午9 04 56](https://user-images.githubusercontent.com/65380823/117540368-a0ad3200-b041-11eb-9e24-0ed0763132bb.png)

when I use ncnn export yolo5s.param, how can I get the result like the first picture instead of the second one? when I use the old version export.py,i can get the first picture. thank for your reply!!

**Comment by YuhangLiu-602** [comment]
> if you wanna use ncnn's yolov5 demo cpp, do not add model.model = model.model[:-1].
> 
> just fix export.py (add training=TrainingMode.TRAINING for torch.onnx.export)
> 
> ```
>      print(f'{prefix} starting export with onnx {onnx.__version__}...')
>      f = opt.weights.replace('.pt', '.onnx')  # filename
> ```
> 
> * ```
>      torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
>                        dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # size(1,3,640,640)
>                                      'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
>   ```

![截屏2021-05-08 下午9 06 19](https://user-images.githubusercontent.com/65380823/117561172-01cb1900-b0c7-11eb-9afa-f89b2c209c60.png)
![截屏2021-05-08 下午9 04 56](https://user-images.githubusercontent.com/65380823/117561175-08f22700-b0c7-11eb-8ad2-c5ed1f565eb4.png)
when I use ncnn export yolo5s.param, how can I get the result like the first picture instead of the second one? when I use the old version export.py,i can get the first picture. thank for your reply!!

**Comment by rikichou** [comment]
just use the newest export.py, and modify parameter of tf.onnx.export():
        torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
                          dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # size(1,3,640,640)
                                        'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)

add training=TrainingMode.TRAINING, and the exported model is in training mode!

**Comment by rikichou** [comment]
> just use the newest export.py, and modify parameter of tf.onnx.export():
> torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
> dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'}, # size(1,3,640,640)
> 'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
> 
> add training=TrainingMode.TRAINING, and the exported model is in training mode!



> > if you wanna use ncnn's yolov5 demo cpp, do not add model.model = model.model[:-1].
> > just fix export.py (add training=TrainingMode.TRAINING for torch.onnx.export)
> > ```
> >      print(f'{prefix} starting export with onnx {onnx.__version__}...')
> >      f = opt.weights.replace('.pt', '.onnx')  # filename
> > ```
> > 
> > 
> > 
> > * ```
> >      torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
> >                        dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # size(1,3,640,640)
> >                                      'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
> >   ```
> 
> ![截屏2021-05-08 下午9 06 19](https://user-images.githubusercontent.com/65380823/117561172-01cb1900-b0c7-11eb-9afa-f89b2c209c60.png)
> ![截屏2021-05-08 下午9 04 56](https://user-images.githubusercontent.com/65380823/117561175-08f22700-b0c7-11eb-8ad2-c5ed1f565eb4.png)
> when I use ncnn export yolo5s.param, how can I get the result like the first picture instead of the second one? when I use the old version export.py,i can get the first picture. thank for your reply!!

just use the newest export.py, and modify parameter of tf.onnx.export():
torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'}, # size(1,3,640,640)
'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)

add training=TrainingMode.TRAINING, and the exported model is in training mode!
do not remember add necessary import

import torch._C as _C
TrainingMode = _C._onnx.TrainingMode

**Comment by YuhangLiu-602** [comment]
> > just use the newest export.py, and modify parameter of tf.onnx.export():
> > torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
> > dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'}, # size(1,3,640,640)
> > 'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
> > add training=TrainingMode.TRAINING, and the exported model is in training mode!
> 
> > > if you wanna use ncnn's yolov5 demo cpp, do not add model.model = model.model[:-1].
> > > just fix export.py (add training=TrainingMode.TRAINING for torch.onnx.export)
> > > ```
> > >      print(f'{prefix} starting export with onnx {onnx.__version__}...')
> > >      f = opt.weights.replace('.pt', '.onnx')  # filename
> > > ```
> > > 
> > > 
> > > 
> > > * ```
> > >      torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
> > >                        dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'},  # size(1,3,640,640)
> > >                                      'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
> > >   ```
> > 
> > 
> > ![截屏2021-05-08 下午9 06 19](https://user-images.githubusercontent.com/65380823/117561172-01cb1900-b0c7-11eb-9afa-f89b2c209c60.png)
> > ![截屏2021-05-08 下午9 04 56](https://user-images.githubusercontent.com/65380823/117561175-08f22700-b0c7-11eb-8ad2-c5ed1f565eb4.png)
> > when I use ncnn export yolo5s.param, how can I get the result like the first picture instead of the second one? when I use the old version export.py,i can get the first picture. thank for your reply!!
> 
> just use the newest export.py, and modify parameter of tf.onnx.export():
> torch.onnx.export(model, img, f, training=TrainingMode.TRAINING, verbose=False, opset_version=12, input_names=['images'],
> dynamic_axes={'images': {0: 'batch', 2: 'height', 3: 'width'}, # size(1,3,640,640)
> 'output': {0: 'batch', 2: 'y', 3: 'x'}} if opt.dynamic else None)
> 
> add training=TrainingMode.TRAINING, and the exported model is in training mode!
> do not remember add necessary import
> 
> import torch._C as _C
> TrainingMode = _C._onnx.TrainingMode

thank for your reply，it works！
